### PR TITLE
feat: Support trigger a recommendation manually

### PR DESCRIPTION
**Pull request by qmhu**
#### What type of PR is this?

Feature

#### What this PR does / why we need it:

Manually triggering a recommendation can be done through the following steps:

1. Obtain the runNumber from the recommendationRule.
image

2. Modify the run number annotation for the recommendation that should be triggered. The run number should be lower than the runNumber in the recommendationRule. For example, if recommendationRule.status.runNumber=6, then change the annotation value to 5.
image

3. The recommendationRule controller checks for updates every 10 seconds. Once it detects the modified run number, it triggers the recommendation flow and updates the recommendationRule.status and recommendation. The recommendation's runNumber is modified to match that of the recommendationRule. You can use this to know when the manual trigger is finished.

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/controller/recommendation/recommendation_rule_controller.go [hunk 1]
```diff
@@ -278,11 +286,108 @@ func (c *RecommendationRuleController) SetupWithManager(mgr ctrl.Manager) error
...
+	}()
+}
+
+func (c *RecommendationRuleController) setStatus(recommendationRule *analysisv1alph1.RecommendationRule, isRunning bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.RecommendationRuleStatus[string(recommendationRule.UID)] = isRunning
+}
+
+func (c *RecommendationRuleController) getStatus(recommendationRule *analysisv1alph1.RecommendationRule) (bool, bool) {
```
Comment (whitebear009) [inline pkg/controller/recommendation/recommendation_rule_controller.go:383]: the same as above

**Comment by whitebear009** [comment]
It seems that only a single recommendation can be manually executed. Is it necessary to support manual execution of the recommendationrule?

File: pkg/controller/recommendation/recommendation_rule_controller.go [hunk 2]
```diff
@@ -278,11 +286,108 @@ func (c *RecommendationRuleController) SetupWithManager(mgr ctrl.Manager) error
...
+					}
+
+				}
+
+			}
+		}
+	}()
+}
+
+func (c *RecommendationRuleController) setStatus(recommendationRule *analysisv1alph1.RecommendationRule, isRunning bool) {
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/controller/recommendation/recommendation_rule_controller.go]: remove it

File: pkg/controller/recommendation/recommendation_rule_controller.go (hunk 1 above)
Comment (qmhu) [reply pkg/controller/recommendation/recommendation_rule_controller.go]: also remove it

**Comment by qmhu** [comment]
> 

Last implement has a limitation: if the recommendation rule controller is reconciling, the manual trigger will wait, this is unexpectable. So we use a trigger controller to watch the decrease event.

How to trigger a recommendation is trick work. We can't modify the spec, but recommendationRule controller only watch spec changes...

**Comment by qmhu** [comment]
@whitebear009 any comments?

File: pkg/controller/recommendation/recommendation_trigger_controller.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+func (h *recommendationEventHandler) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (h *recommendationEventHandler) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	newRecommendation := evt.ObjectNew.(*analysisv1alpha1.Recommendation)
+	oldRecommendation := evt.ObjectOld.(*analysisv1alpha1.Recommendation)
+	klog.V(6).Infof("recommendation %s OnUpdate", klog.KObj(newRecommendation))
+
+	oldRunNumber, _ := utils.GetRunNumber(oldRecommendation)
+	newRunNumber, _ := utils.GetRunNumber(newRecommendation)
```
Comment (whitebear009) [inline pkg/controller/recommendation/recommendation_trigger_controller.go:152]: the error is not handled

Comment (qmhu) [reply pkg/controller/recommendation/recommendation_trigger_controller.go:152]: if get run number failed , we treat it as a empty number which by default is 0.

**Pull request merged**